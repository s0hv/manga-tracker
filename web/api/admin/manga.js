const express = require('express');
const { body, param } = require('express-validator');
const { QueryResultError } = require('pg-promise').errors;
const { updateMangaTitle } = require('../../db/admin/manga');
const { requiresUser } = require('../../db/auth');
const { handleError } = require('../../db/utils');
const {
  scheduleMangaRun,
  getScheduledRuns,
  deleteScheduledRun,
} = require('../../db/admin/management');
const {
  validateAdminUser,
  mangaIdValidation,
  serviceIdValidation,
  handleValidationErrors,
} = require('../../utils/validators');


module.exports = () => {
  const router = express.Router();
  router.use(requiresUser);

  router.use('/:mangaId(\\d+)/scheduledRuns', [
    validateAdminUser(),
    mangaIdValidation(param('mangaId')),
  ]);
  router.get('/:mangaId(\\d+)/scheduledRuns', handleValidationErrors, (req, res) => {
    getScheduledRuns(req.params.mangaId)
      .then(rows => {
        res.status(200).json({
          data: rows,
        });
      })
      .catch(err => handleError(err, res));
  });

  const scheduleRunUrl = '/:mangaId(\\d+)/scheduledRun/:serviceId(\\d+)';
  router.use(scheduleRunUrl, [
    validateAdminUser(),
    mangaIdValidation(param('mangaId')),
    serviceIdValidation(param('serviceId')),
    handleValidationErrors,
  ]);
  router.route(scheduleRunUrl)
    .post((req, res) => {
      scheduleMangaRun(req.params.mangaId, req.params.serviceId, req.user.userId)
        .then(row => {
          res.status(200).json({
            inserted: row,
          });
        })
        .catch(err => handleError(err, res));
    })

    .delete((req, res) => {
      deleteScheduledRun(req.params.mangaId, req.params.serviceId)
        .then(rows => {
          if (rows.rowCount > 0) {
            res.status(200).end();
          } else {
            res.status(404).end();
          }
        })
        .catch(err => handleError(err, res));
    });

  const updateTitleUrl = '/:mangaId(\\d+)/title';
  router.use(updateTitleUrl, [
    validateAdminUser(),
    mangaIdValidation(param('mangaId')),
    body('title').isString().bail().isLength({ min: 1 }),
  ]);
  router.post(updateTitleUrl, handleValidationErrors, (req, res) => {
    updateMangaTitle(req.params.mangaId, req.body.title)
      .then(row => {
        const msg = row ?
          `Replaced old alias with current title "${row.title}"` :
          `Alias not found. Scrapping old title`;
        res.json({ message: msg });
      })
      .catch(err => {
        if (err instanceof QueryResultError) {
          res.status(err.status).json({ error: 'Manga not found' });
          return;
        }
        handleError(err, res);
      });
  });

  return router;
};
