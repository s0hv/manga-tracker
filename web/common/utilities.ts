import { Group } from '@/webUtils/utilities';

type GetKey<T> = (value: T) => string;
interface GroupByOptions<B extends boolean = boolean, C extends boolean = boolean> {
  keepOrder: B
  returnAsDict?: C
}

type GroupBy = {
  <T, B extends true = true, C extends false = false>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): Group<T>[]
  <T, B extends false = false, C extends false = false>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): T[][]
  <T, B extends false = false, C extends true = true>(arr: T[], getKeyOrKey: keyof T | GetKey<T>, options?: GroupByOptions<B, C>): {[key: string]: T[]}
}

/**
 *
 * @param arr
 * @param getKeyOrKey Either an object property name or a function that returns the group key
 * @param options
 * @param options.keepOrder If true the order of the array will be kept the same
 * meaning the same key can be grouped multiple times
 * @param options.returnAsDict If true returns the grouped values as a dictionary. Ignored if keepOrder is true.
 */
export const groupBy: GroupBy = <T, >(
  arr: T[],
  getKeyOrKey: keyof T | GetKey<T>,
  { keepOrder = true, returnAsDict = false } = {} as GroupByOptions<true, false>
  // By adding "| T" to the return type somehow fixes overload errors WTF???
  // GJ typescript
): Group<T>[] | T[][] | T | {[key: string]: T[]} => {
  if (!Array.isArray(arr)) {
    throw new TypeError('Input must be an array');
  }

  if (arr.length === 0) {
    return returnAsDict ? {} : [];
  }

  let getKey: GetKey<T>;
  if (typeof getKeyOrKey === 'string') {
    const _key = getKeyOrKey;
    getKey = (o) => (o as any)[_key];
  } else {
    getKey = getKeyOrKey as any;
  }

  if (keepOrder) {
    const groups: Group<T>[] = [];
    let currentGroup: Group<T> = {
      group: '',
      arr: [],
    };

    let currentGroupKey: string;
    arr.forEach((o) => {
      const key = getKey(o);
      if (key !== currentGroupKey) {
        if (currentGroup.arr.length > 0) {
          groups.push(currentGroup);
        }
        currentGroup = {
          group: '',
          arr: [],
        };
        currentGroup.group = key;
        currentGroupKey = key;
      }

      currentGroup.arr.push(o);
    });

    if (currentGroup.arr.length > 0) {
      currentGroup.group = currentGroupKey!;
      groups.push(currentGroup);
    }

    return groups;
  }

  // Set is iterated in insertion order https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#description
  const order: Set<string> = new Set();
  const group: Record<string, T[]> = {};

  arr.forEach((o) => {
    const key = getKey(o);
    // If group made add to that
    if (order.has(key)) {
      group[key].push(o);
    } else {
      // No group yet. Create new group and add group key to order
      group[key] = [o];
      order.add(key);
    }
  });

  if (returnAsDict) {
    return group;
  }

  const retVal: T[][] = [];
  order.forEach((groupKey) => retVal.push(group[groupKey]));
  return retVal;
};
