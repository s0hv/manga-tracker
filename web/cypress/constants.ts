export const nextAuthSessionCookie = 'next-auth.session-token';
