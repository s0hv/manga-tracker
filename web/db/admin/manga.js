import { db } from '..';

export const updateMangaTitle = (mangaId, newTitle) => {
  const sql = `UPDATE manga
               SET title=$1
               WHERE manga_id=$2
               RETURNING (SELECT title FROM manga WHERE manga_id=$2)`;

  const aliasSql = `UPDATE manga_alias
                    SET title=$1
                    WHERE manga_id=$2 AND title=$3 AND NOT EXISTS(SELECT 1 FROM manga_alias WHERE manga_id=$2 AND title=$1)
                    RETURNING title`;

  return db.one(sql, [newTitle, mangaId])
    .then(row => db.oneOrNone(aliasSql, [row.title, mangaId, newTitle]));
};
