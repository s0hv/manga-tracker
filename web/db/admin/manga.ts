import snakecaseKeys from 'snakecase-keys';
import { db } from '..';
import type { DatabaseId, MangaId, MangaInfoUpdate } from '@/types/dbTypes';
import type { MangaService, MangaServiceUpdateData } from '@/types/api/manga';
import { generateUpdate } from '../utils';
import { NoColumnsError } from '../errors';
import { BadRequest } from '../../utils/errors';

export const updateMangaTitle = (mangaId: MangaId, newTitle: string) => {
  const sql = `UPDATE manga
               SET title=$1
               WHERE manga_id=$2
               RETURNING (SELECT title FROM manga WHERE manga_id=$2)`;

  const aliasSql = `UPDATE manga_alias
                    SET title=$1
                    WHERE manga_id=$2 AND title=$3 AND NOT EXISTS(SELECT 1 FROM manga_alias WHERE manga_id=$2 AND title=$1)
                    RETURNING title`;

  return db.one(sql, [newTitle, mangaId])
    .then(row => db.oneOrNone(aliasSql, [row.title, mangaId, newTitle]));
};

export const updateMangaInfo = (mangaInfo: MangaInfoUpdate) => {
  const sql = 'UPDATE manga_info SET status=$1 WHERE manga_id=$2';
  return db.result(sql, [mangaInfo.status, mangaInfo.mangaId]);
};

export const getMangaServices = (mangaId: MangaId): Promise<MangaService[]> => {
  const sql = `SELECT * FROM manga_service WHERE manga_id=$1`;

  return db.manyOrNone<MangaService>(sql, [mangaId]);
};


export const updateMangaService = async (
  mangaId: MangaId, serviceId: DatabaseId,
  { disabled, nextUpdate }: MangaServiceUpdateData
) => {
  let update: string;
  try {
    update = generateUpdate(snakecaseKeys({ disabled, nextUpdate }), 'manga_service');
  } catch (e) {
    if (e instanceof NoColumnsError) {
      throw new BadRequest('No valid columns given to update');
    }

    throw e;
  }
  const sql = `${update} WHERE manga_id=$1 AND service_id=$2`;

  return db.result(sql, [mangaId, serviceId]);
};
