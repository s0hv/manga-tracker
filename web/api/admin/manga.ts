import express from 'express';
import { body, param } from 'express-validator';
import pgPromise from 'pg-promise';
import {
  createMangaService,
  getMangaServices,
  updateMangaInfo,
  updateMangaService,
  updateMangaTitle,
} from '@/db/admin/manga';
import { requiresUser } from '@/db/auth';
import { handleError } from '@/db/utils';
import {
  deleteScheduledRun,
  getScheduledRuns,
  scheduleMangaRun,
} from '@/db/admin/management';
import {
  databaseIdValidation,
  handleValidationErrors,
  mangaIdValidation,
  serviceIdValidation,
  validateAdminUser,
} from '../../utils/validators.js';
import { getMangaForElastic } from '@/db/manga';
import { updateManga } from '@/db/elasticsearch/manga';
import { MangaStatus } from '@/types/dbTypes';

const { QueryResultError } = pgPromise.errors;

export default () => {
  const router = express.Router();
  router.use([requiresUser, validateAdminUser()]);

  router.use('/:mangaId(\\d+)/scheduledRuns', [
    mangaIdValidation(param('mangaId')),
  ]);
  router.get('/:mangaId(\\d+)/scheduledRuns', handleValidationErrors, (req, res) => {
    getScheduledRuns(req.params.mangaId)
      .then(rows => {
        res.status(200).json({
          data: rows,
        });
      })
      .catch(err => handleError(err, res));
  });

  const scheduleRunUrl = '/:mangaId(\\d+)/scheduledRun/:serviceId(\\d+)';
  router.use(scheduleRunUrl, [
    mangaIdValidation(param('mangaId')),
    serviceIdValidation(param('serviceId')),
    handleValidationErrors,
  ]);
  router.route(scheduleRunUrl)
    .post((req, res) => {
      scheduleMangaRun(req.params.mangaId, req.params.serviceId, req.user.userId)
        .then(row => {
          res.status(200).json({
            inserted: row,
          });
        })
        .catch(err => handleError(err, res));
    })

    .delete((req, res) => {
      deleteScheduledRun(req.params.mangaId, req.params.serviceId)
        .then(rows => {
          if (rows.rowCount > 0) {
            res.status(200).end();
          } else {
            res.status(404).end();
          }
        })
        .catch(err => handleError(err, res));
    });

  const updateTitleUrl = '/:mangaId(\\d+)/title';
  router.use(updateTitleUrl, [
    mangaIdValidation(param('mangaId')),
    body('title').isString().bail().isLength({ min: 1 }),
  ]);
  router.post(updateTitleUrl, handleValidationErrors, (req, res) => {
    updateMangaTitle(req.params.mangaId, req.body.title)
      .then(row => {
        getMangaForElastic(req.params.mangaId)
          .then(manga => updateManga(manga.mangaId, manga))
          .finally(() => {
            const msg = row ?
              `Replaced old alias with current title "${row.title}"` :
              `Alias not found. Scrapping old title`;
            res.json({ message: msg });
          });
      })
      .catch(err => {
        if (err instanceof QueryResultError) {
          res.status(404).json({ error: 'Manga not found' });
          return;
        }
        handleError(err, res);
      });
  });

  const updateInfoPath = '/:mangaId(\\d+)/info';
  router.post(updateInfoPath, ...[
    mangaIdValidation(param('mangaId')),
    body('status').isInt({ min: MangaStatus.ONGOING, max: MangaStatus.HIATUS }).toInt(),
    handleValidationErrors,
  ], (req, res) => {
    updateMangaInfo({
      status: req.body.status,
      mangaId: req.params!.mangaId,
    })
      .then(r => {
        if (r.rowCount === 0) return res.sendStatus(404);

        res.sendStatus(200);
      })
      .catch(err => handleError(err, res));
  });

  router.get('/:mangaId(\\d+)/services', ...[
    mangaIdValidation(param('mangaId')),
    handleValidationErrors,
  ], (req, res) => {
    getMangaServices(req.params!.mangaId)
      .then(r => {
        if (r.length === 0) return res.sendStatus(404);

        res.json(r);
      })
      .catch(err => handleError(err, res));
  });

  router.post('/:mangaId(\\d+)/services/:serviceId(\\d+)', ...[
    mangaIdValidation(param('mangaId')),
    databaseIdValidation(param('serviceId')),
    body('mangaService')
      .isObject({ strict: true }),
    body('mangaService.disabled')
      .optional()
      .isBoolean({ strict: true }),
    body('mangaService.nextUpdate')
      .optional({ nullable: true })
      .isISO8601({ strict: true })
      .toDate(),
    handleValidationErrors,
  ], (req, res) => {
    updateMangaService(req.params!.mangaId, req.params!.serviceId, req.body.mangaService)
      .then(r => {
        if (r.rowCount === 0) return res.sendStatus(404);

        res.sendStatus(200);
      })
      .catch(err => handleError(err, res));
  });

  router.post('/:mangaId(\\d+)/services/:serviceId(\\d+)/create', ...[
    mangaIdValidation(param('mangaId')),
    databaseIdValidation(param('serviceId')),
    body('mangaService')
      .isObject({ strict: true }),
    body('mangaService.titleId')
      .isString(),
    body('mangaService.feedUrl')
      .optional({ nullable: true })
      .isString(),
    handleValidationErrors,
  ], (req, res) => {
    createMangaService(req.params!.mangaId, req.params!.serviceId, req.body.mangaService)
      .then(r => {
        if (r.rowCount === 0) return res.sendStatus(404);

        res.sendStatus(200);
      })
      .catch(err => handleError(err, res));
  });

  return router;
};
