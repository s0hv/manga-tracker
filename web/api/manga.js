import { query, param } from 'express-validator';

import { db } from '../db';
import { handleError } from '../db/utils.js';
import { getChapters } from '../db/chapter.js';
import { getOptionalNumberParam } from '../utils/utilities.js';
import {
  mangaIdValidation,
  hadValidationError,
  validateAdminUser,
  databaseIdValidation,
  handleValidationErrors,
} from '../utils/validators.js';
import { requiresUser } from '../db/auth.js';
import { getFullManga } from '../db/manga.js';

const BASE_URL = '/api/manga';

export default app => {
  app.post('/api/manga/merge', requiresUser, [
    validateAdminUser(),
    databaseIdValidation(query('base')),
    databaseIdValidation(query('toMerge')),
    databaseIdValidation(query('service')).optional(),
  ], (req, res) => {
    if (hadValidationError(req, res)) return;

    if (req.query.base === req.query.toMerge) {
      res.status(400).json({ error: 'Given ids are equal' });
      return;
    }

    const sql = 'SELECT * FROM merge_manga($1, $2, $3)';
    db.oneOrNone(sql, [req.query.base, req.query.toMerge, req.query.service || null])
      .then(row => {
        if (!row) {
          return res.status(500).json({ error: 'No modifications done' });
        }
        res.status(200).json(row);
      })
      .catch(err => handleError(err, res));
  });

  /**
   * @openapi
   * /manga/{mangaId}:
   *   get:
   *     summary: Get manga
   *     description: Get information about a manga
   *     parameters:
   *       - name: mangaId
   *         in: path
   *         required: true
   *         description: Id of the manga
   *         schema:
   *           $ref: '#/components/schemas/databaseId'
   *
   *     responses:
   *       200:
   *         description: Returns a JSON containing information about the manga
   *         content:
   *           application/json:
   *             schema:
   *               type: object
   *               properties:
   *                 data:
   *                   $ref: '#/components/schemas/fullManga'
   *               required:
   *                 - data
   *       400:
   *         $ref: '#/components/responses/validationError'
   *       404:
   *         $ref: '#/components/responses/notFound'
   */
  app.get('/api/manga/:mangaId', [
    mangaIdValidation(param('mangaId')),
    handleValidationErrors,
  ], (req, res) => {
    getFullManga(req.params.mangaId)
      .then(manga => {
        if (!manga) {
          res.status(404).json({ error: 'Manga not found' });
          return;
        }
        res.json({ data: manga });
      })
      .catch(err => handleError(err, res));
  });

  /**
   *  @openapi
   *  /manga/{mangaId}/chapters:
   *    get:
   *      summary: Get chapters
   *      description: >
   *        Get the chapters of a manga. Results are sorted by the release date
   *        in descending order.
   *      parameters:
   *        - name: mangaId
   *          in: path
   *          required: true
   *          description: Id of the manga
   *          schema:
   *            $ref: '#/components/schemas/databaseId'
   *
   *        - name: limit
   *          in: query
   *          required: false
   *          description: Amount of chapters to fetch.
   *          schema:
   *            type: integer
   *            minimum: 0
   *            maximum: 200
   *            default: 50
   *
   *        - name: offset
   *          in: query
   *          required: false
   *          description: Offset used when fetching.
   *          schema:
   *            type: integer
   *            minimum: 0
   *            default: 0
   *
   *        - name: sortBy
   *          in: query
   *          required: false
   *          description: The column used for sorting.
   *          schema:
   *            type: string
   *            enum: ['chapter_number', 'group', 'chapter_id', 'release_date']
   *
   *        - name: sort
   *          in: query
   *          required: false
   *          description: The sorting direction.
   *          schema:
   *            type: string
   *            enum: ['asc', 'desc']
   *
   *      responses:
   *        200:
   *          description: Returns a list of chapters and the amount of all chapters
   *          content:
   *            application/json:
   *              schema:
   *                type: object
   *                properties:
   *                  data:
   *                    type: object
   *                    properties:
   *                      count:
   *                        type: integer
   *                        minimum: 0
   *                      chapters:
   *                        type: array
   *                        items:
   *                          $ref: '#/components/schemas/chapter'
   *
   *                    required:
   *                      - count
   *                      - chapters
   *
   *                required:
   *                  - data
   *        400:
   *          $ref: '#/components/responses/validationError'
   *        404:
   *          $ref: '#/components/responses/notFound'
   */
  app.get(`${BASE_URL}/:mangaId(\\d+)/chapters`, [
    mangaIdValidation(param('mangaId')),
    query('limit').isInt({ min: 0, max: 200 }).optional().withMessage('Limit must be an integer between 0 and 200'),
    query('offset').isInt({ min: 0 }).optional().withMessage('Offset must be a positive integer'),
    query('sortBy')
      .isString()
      .optional()
      .bail()
      .toLowerCase()
      .isIn(['chapter_id', 'chapter_number', 'release_date', 'group'])
      .withMessage('Sort column must be one of "chapter_id", "chapter_number", "release_date", "group"'),
    query('sort')
      .isString()
      .optional()
      .bail()
      .toLowerCase()
      .isIn(['asc', 'desc'])
      .withMessage('Sorting direction must be one of "asc" or "desc"'),
    handleValidationErrors,
  ], (req, res) => {
    const mangaId = Number(req.params.mangaId);
    let limit;
    let offset;

    try {
      limit = getOptionalNumberParam(req.query.limit, 50);
      offset = getOptionalNumberParam(req.query.offset, 0);
    } catch (err) {
      if (err instanceof TypeError) {
        res.status(400).json({ error: err.message });
      } else {
        res.status(500).json({ error: 'Internal server error' });
      }
      return;
    }

    const sortBy = [];
    const isDesc = req.query.sort === 'desc';

    if (req.query.sortBy === 'chapter_number') {
      sortBy.push({
        col: 'chapter_number',
        desc: isDesc,
      });
      sortBy.push({
        col: 'chapter_decimal',
        desc: isDesc,
        nullsLast: isDesc,
      });
    } else if (req.query.sortBy) {
      sortBy.push({
        col: req.query.sortBy,
        desc: isDesc,
        nullsLast: isDesc,
      });
    }

    getChapters(mangaId, limit, offset, sortBy)
      .then(rows => {
        if (!rows) {
          res.status(404).json({ error: 'No manga found with given id' });
          return;
        }
        res.json({ data: rows });
      })
      .catch(err => handleError(err, res));
  });
};
