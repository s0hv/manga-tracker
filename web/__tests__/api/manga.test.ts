import request from 'supertest';
import { csrfMissing } from '../../utils/constants';
import { userForbidden, userUnauthorized } from '../constants';

import initServer from '../initServer';
import stopServer from '../stopServer';
import {
  adminUser,
  configureJestOpenAPI,
  expectErrorMessage,
  getErrorMessage,
  normalUser,
  withUser,
} from '../utils';
import { apiRequiresAdminUserPostTests } from './utilities';
import type { DatabaseId, MangaId } from '@/types/dbTypes';
import { createMangaService } from '../dbutils';
import { deleteManga, updateManga } from '@/db/elasticsearch/manga';
import { getMangaPartial } from '@/db/manga';
import { NoResultsError } from '@/db/errors';

let httpServer: any;
const serverReference = {
  httpServer,
};

jest.mock('../../db/elasticsearch/manga', () => {
  const original = jest.requireActual('../../db/elasticsearch/manga'); // Step 2.
  return {
    ...jest.requireActual('../../db/elasticsearch/manga'),
    updateManga: jest.fn().mockImplementation(original.updateManga),
    deleteManga: jest.fn().mockImplementation(original.deleteManga),
  };
});

beforeAll(async () => {
  ({ httpServer } = await initServer());
  serverReference.httpServer = httpServer;
  await configureJestOpenAPI();
});

afterAll(async () => {
  await stopServer(httpServer);
});

describe('POST /api/manga/merge', () => {
  it('Returns 403 without CSRF token', async () => {
    await request(httpServer)
      .post('/api/manga/merge')
      .expect(403)
      .expect(expectErrorMessage(csrfMissing));
  });

  it('Returns 401 without user authentication', async () => {
    await request(httpServer)
      .post('/api/manga/merge')
      .csrf()
      .expect(401)
      .expect(expectErrorMessage(userUnauthorized));
  });

  it('Returns 403 without admin rights', async () => {
    await withUser(normalUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge')
        .csrf()
        .expect(403)
        .expect(expectErrorMessage(userForbidden));
    });
  });

  it('Returns 400 with invalid base id', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?base=-1')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('-1', 'base'));

      await request(httpServer)
        .post('/api/manga/merge?base=NaN')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('NaN', 'base'));

      await request(httpServer)
        .post('/api/manga/merge?base=abc')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('abc', 'base'));
    });
  });

  it('Returns 400 with invalid toMerge id', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?toMerge=-1')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('-1', 'toMerge'));

      await request(httpServer)
        .post('/api/manga/merge?toMerge=NaN')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('NaN', 'toMerge'));

      await request(httpServer)
        .post('/api/manga/merge?toMerge=abc')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('abc', 'toMerge'));
    });
  });

  it('Returns 400 when base equals toMerge', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?toMerge=5&base=5')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('Given ids are equal'));
    });
  });

  it('Returns 400 with invalid service id', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?toMerge=1&base=2&service=-1')
        .csrf()
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());

      await request(httpServer)
        .post('/api/manga/merge?toMerge=1&base=2&service=abc')
        .csrf()
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());
    });
  });
});


describe('GET /api/manga/:mangaId', () => {
  it('returns 404 with non existent manga id', async () => {
    await request(httpServer)
      .get(`/api/manga/9999999`)
      .expect(404)
      .satisfiesApiSpec()
      .expect(expectErrorMessage('Manga not found'));
  });

  it('returns 400 with non positive integers', async () => {
    await request(httpServer)
      .get(`/api/manga/-1`)
      .expect(400)
      .satisfiesApiSpec()
      .expect(expectErrorMessage('-1', 'mangaId', /positive integer/));

    await request(httpServer)
      .get(`/api/manga/abc`)
      .expect(400)
      .satisfiesApiSpec()
      .expect(expectErrorMessage('abc', 'mangaId', /positive integer/));
  });

  it('returns 200 with valid manga id', async () => {
    await request(httpServer)
      .get(`/api/manga/1`)
      .expect(200)
      .satisfiesApiSpec();
  });
});

const getChapterCount = (body: any) => (body && body?.data?.chapters?.length) || 0;

describe('GET /api/manga/:mangaId/chapters', () => {
  const validUrl = '/api/manga/1/chapters';

  it('returns 404 with invalid manga id', async () => {
    await request(httpServer)
      .get(`/api/manga/9999999/chapters`)
      .expect(404);

    await request(httpServer)
      .get(`/api/manga/-1/chapters`)
      .expect(404);
  });

  it('returns 400 with an invalid limit', async () => {
    const errorMessage = 'Limit must be an integer between 0 and 200';
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?limit=invalid`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('invalid', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=NaN`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('NaN', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=Infinity`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('Infinity', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=-1`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('-1', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=300`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('300', 'limit', errorMessage))
        .expect(400),
    ]);
  });

  it('Returns chapters with valid limit', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}`)
        .expect('Content-Type', /json/)
        .satisfiesApiSpec()
        .expect(200),

      request(httpServer)
        .get(`${validUrl}?limit=5`)
        .expect('Content-Type', /json/)
        .expect(200)
        .satisfiesApiSpec()
        .expect(res => expect(getChapterCount(res.body)).toBe(5)),
    ]);
  });

  it('Returns 400 with invalid offset', async () => {
    const errorMessage = 'Offset must be a positive integer';

    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?offset=NaN`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('NaN', 'offset', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?offset=-1`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('-1', 'offset', errorMessage))
        .expect(400),
    ]);
  });

  it('Returns chapters with valid offset', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?offset=3`)
        .satisfiesApiSpec()
        .expect('Content-Type', /json/)
        .expect(200),
    ]);
  });

  it('Returns 400 with invalid sort column', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?sortBy=test`)
        .expect('Content-Type', /json/)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot())
        .expect(400),
    ]);
  });

  it('Returns 400 with invalid sort direction', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?sort=test`)
        .expect('Content-Type', /json/)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot())
        .expect(400),
    ]);
  });

  it('Returns chapters with valid sort column and direction', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?sortBy=chapter_number&sort=desc`)
        .satisfiesApiSpec()
        .expect('Content-Type', /json/)
        .expect(200),
    ]);
  });
});

describe('POST /api/manga/merge', () => {
  const getUrl = (base: MangaId, toMerge: MangaId, service?: DatabaseId) => {
    return `/api/manga/merge?base=${base}&toMerge=${toMerge}${service ? `&service=${service}` : ''}`;
  };

  apiRequiresAdminUserPostTests(serverReference, getUrl(1, 2));

  const esDeleteMock = jest.fn();
  const esUpdateMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    esDeleteMock.mockResolvedValue(undefined);
    esUpdateMock.mockResolvedValue(undefined);
    (deleteManga as jest.Mock).mockImplementation(esDeleteMock);
    (updateManga as jest.Mock).mockImplementation(esUpdateMock);
  });

  it('Returns 400 with invalid params', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(getUrl('a', 1))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('a', 'base'));

      await request(httpServer)
        .post(getUrl(1, 'x'))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('x', 'toMerge'));

      await request(httpServer)
        .post(getUrl(1, 1, 'not int'))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('not int', 'service'));

      await request(httpServer)
        .post(getUrl(1.1, 2))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('1.1', 'base'));
    });
  });

  it('Returns 400 with same manga ids', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(getUrl(10, 10))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('Given ids are equal'));
    });
  });

  it('Returns 400 when manga is missing', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(getUrl(9999, 99999))
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('Not null value was null'));
    });
  });

  it('Returns 200 when manga is found', async () => {
    const m1 = await createMangaService(1);
    const m2 = await createMangaService(2);

    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(getUrl(m1, m2))
        .csrf()
        .expect(200)
        .expect(res => expect(res.body).toEqual({ aliasCount: 0, chapterCount: 0 }));
    });

    expect(esUpdateMock).toHaveBeenCalledOnce();
    expect(esUpdateMock).toHaveBeenLastCalledWith(m1, expect.objectContaining({
      mangaId: m1,
    }));

    expect(esDeleteMock).toHaveBeenCalledOnce();
    expect(esDeleteMock).toHaveBeenLastCalledWith(m2.toString());

    // Make sure the other manga is gone from the db
    await expect(getMangaPartial(m2))
      .rejects
      .toThrow(NoResultsError);
  });

  it('Returns 200 when manga is found with specific service', async () => {
    const m1 = await createMangaService(1);
    const serviceId = 2;
    const m2 = await createMangaService(serviceId);
    await createMangaService(3, m2);

    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(getUrl(m1, m2, serviceId))
        .csrf()
        .expect(200)
        .expect(res => expect(res.body).toEqual({ aliasCount: 0, chapterCount: 0 }));
    });

    expect(esDeleteMock).not.toHaveBeenCalled();

    let loops = 0;
    const waitForUpdates = async (): Promise<void> => {
      if (esUpdateMock.mock.calls.length >= 2) return;

      loops++;
      // Shouldn't take too long so just quit early and save a few seconds
      if (loops > 4) return;

      await new Promise(resolve => setTimeout(resolve, 50));
      return waitForUpdates();
    };

    await waitForUpdates();

    expect(esUpdateMock).toHaveBeenCalledTimes(2);
    expect(esUpdateMock).toHaveBeenCalledWith(m2, expect.objectContaining({
      mangaId: m2,
    }));
    expect(esUpdateMock).toHaveBeenCalledWith(m1, expect.objectContaining({
      mangaId: m1,
    }));

    // Make sure the other manga stays in the manga
    expect(await getMangaPartial(m2)).toBeDefined();
  });
});
