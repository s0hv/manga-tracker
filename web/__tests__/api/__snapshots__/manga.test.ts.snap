// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/manga/:mangaId/chapters Returns 400 with invalid sort column 1`] = `
Array [
  Object {
    "location": "query",
    "msg": "Sort column must be one of \\"chapter_id\\", \\"chapter_number\\", \\"release_date\\", \\"group\\"",
    "param": "sortBy",
    "value": "test",
  },
]
`;

exports[`GET /api/manga/:mangaId/chapters Returns 400 with invalid sort direction 1`] = `
Array [
  Object {
    "location": "query",
    "msg": "Sorting direction must be one of \\"asc\\" or \\"desc\\"",
    "param": "sort",
    "value": "test",
  },
]
`;

exports[`POST /api/manga/merge Returns 400 with invalid service id 1`] = `
Array [
  Object {
    "location": "query",
    "msg": "Invalid value",
    "param": "service",
    "value": "-1",
  },
]
`;

exports[`POST /api/manga/merge Returns 400 with invalid service id 2`] = `
Array [
  Object {
    "location": "query",
    "msg": "Invalid value",
    "param": "service",
    "value": "abc",
  },
]
`;
